### PR TITLE
SIMP-5125 Create New IMA module

### PR DESCRIPTION
**Pull request by m-morrone**
Copied TPM package and pulled out everything not IMA

SIMP-5125 #close

### CODE REVIEW

**Comment by m-morrone** [comment]
Correct commit message.  All commits can be squashed into a single merge.

File: README.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+[![License](http://img.shields.io/:license-apache-blue.svg)](http://www.apache.org/licenses/LICENSE-2.0.html) [![Build Status](https://travis-ci.org/simp/pupmod-simp-tpm.svg)](https://travis-ci.org/simp/pupmod-simp-ima)
```
Comment (trevor-vaughan) [inline README.md]: These need to be changed to the new format which can be found at the head of `pupmod-simp-simp` and all instances of `tpm` need to be changed to `ima`

File: README.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+5. [Limitations - OS compatibility, etc.](#limitations)
+6. [Development - Guide for contributing to the module](#development)
+    * [Acceptance Tests - Beaker env variables](#acceptance-tests)
+
+
+## Description
+
+This module manages the [Integrity Management Architecture (IMA)](https://sourceforge.net/p/linux-ima/wiki/Home/),
+a tool that verifies the integrity of the system, based on filesystem 
+and file hashes. The IMA class sets up IMA kernel boot flags if
```
Comment (trevor-vaughan) [inline README.md]: This section needs to be updated specifically for IMA

File: build/rpm_metadata/requires [hunk 3]
```diff
@@ -0,0 +1,10 @@
+Requires: pupmod-herculesteam-augeasproviders >= 2.1.3
```
Comment (trevor-vaughan) [inline build/rpm_metadata/requires]: This file should be removed.

File: files/19_linux_tboot_pretxt.diff [hunk 4]
```diff
@@ -0,0 +1,38 @@
+--- 20_linux_tboot	2017-08-02 16:18:41.872143371 -0400
```
Comment (trevor-vaughan) [inline files/19_linux_tboot_pretxt.diff]: This is a diff file that should not be included in the PR.

File: files/20_linux_tboot.diff [hunk 5]
```diff
@@ -0,0 +1,57 @@
+--- 20_linux_tboot	2017-08-04 12:19:31.825655186 -0400
```
Comment (trevor-vaughan) [inline files/20_linux_tboot.diff]: Remove diff files please.

File: manifests/appraise.pp [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+  Simplib::PackageEnsure $package_ensure = $::ima::package_ensure,
+  Boolean                $enable         = true,
+  Stdlib::AbsolutePath   $relabel_file   = "${facts['puppet_vardir']}/simp/.ima_relabel",
+  Stdlib::AbsolutePath   $scriptdir      = '/usr/local/bin',
+  Boolean                $force_fixmode  = false,
+){
+
+  if $enable {
+    # Provides ability to check for special attributes
+    package { 'attr':
```
Comment (trevor-vaughan) [inline manifests/appraise.pp]: This needs to use `ensure_packages` since this is a common package.

File: manifests/appraise.pp [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    package { 'ima-evm-utils':
+      ensure => $package_ensure
+    }
+
+    kernel_parameter { 'ima_appraise_tcb':
+      notify   => Reboot_notify['ima_appraise_reboot'],
+      bootmode => 'normal'
+    }
+
+    kernel_parameter { 'rootflags':
```
Comment (trevor-vaughan) [inline manifests/appraise.pp]: This may cause issues in the future, but I'm not really sure what we can do about it right now.

File: metadata.json [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    {
+      "name": "simp/rsync",
+      "version_requirement": ">= 6.0.0 < 7.0.0"
+    },
+    {
+      "name": "puppetlabs/stdlib",
+      "version_requirement": ">= 4.13.1 < 5.0.0"
+    },
+    {
+      "name": "puppet/yum",
```
Comment (trevor-vaughan) [inline metadata.json]: Not sure why this is required

File: metadata.json [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    {
+      "name": "herculesteam/augeasproviders_grub",
+      "version_requirement": ">= 2.3.0 < 4.0.0"
+    },
+    {
+      "name": "simp/simplib",
+      "version_requirement": ">= 3.1.0 < 4.0.0"
+    },
+    {
+      "name": "simp/rsync",
```
Comment (trevor-vaughan) [inline metadata.json]: Not sure why this is required

File: manifests/appraise.pp [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+#   The directory to place scripts.
+#
+# @param force_fixmode
+#   This will force the system into ``fix_mode`` so you can update files and
+#   then relabel the system - requires a reboot.
+#
+# @author SIMP Team  <https://simp-project.com/>
+#
+class ima::appraise(
+  Simplib::PackageEnsure $package_ensure = $::ima::package_ensure,
```
Comment (lnemsick-simp) [inline manifests/appraise.pp]: This parameter does not exist in the ima class.  
Why is this class public?  You could have resource ordering issues if it is used outside of the ima module, but requires a parameter from the ima class

File: manifests/appraise/relabel.pp [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+#  resources for setting the system into enforce mode.
+#
+#  @param relabel_file   The location of the file that
+#     that indicates a labeling of the file system is needed.
+#
+class ima::appraise::relabel(
+  Stdlib::AbsolutePath  $relabel_file,
+  Stdlib::AbsolutePath  $scriptdir = $ima::appraise::scriptdir,
+){
+
```
Comment (lnemsick-simp) [inline manifests/appraise/relabel.pp]: Should this be private?  It is using ima::appraise::scriptdir, but does not ensure the ima class precedes it in the catalogue

Comment (op-ct) [reply manifests/appraise/relabel.pp]: @lnemsick-simp yes, this class should be private and {{ima::appaise}} should be classified separately from {{ima}} as an unambiguously public class.  That's beyond to the scope of this PR, so I've submitted [SIMP-5102](https://simp-project.atlassian.net/browse/SIMP-5102) as a follow-on issue to handle that.

File: manifests/init.pp [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+#   The size of ``/sys/kernel/security/ima/ascii_runtime_measurements``, in
+#   bytes, that will cause a reboot notification will be sent to the user.
+#
+# @param ensure_packages
+#   Ensure setting for all packages installed by this module
+#
+class ima (
+  Boolean                $enable          = true,
+  Boolean                $manage_appraise = false,
+  Boolean                $manage_policy   = false,
```
Comment (op-ct) [inline manifests/init.pp]: A structural review with @m-morrone today concluded that it would be better to remove `$manage_appraise` and `$manage_policy` from this class so users can classify and set parameters for `ima::appraise` and `ima::policy` directly as unambiguously public classes.  This will simplify the module's structure and help clear up the spec testing issues for Puppet 5.

That work is beyond to the scope of this PR, so I've submitted [SIMP-5102](https://simp-project.atlassian.net/browse/SIMP-5102) as a follow-on to handle it.
